### PR TITLE
morphism contribution

### PR DESCRIPTION
**Pull request by tikoh-station**
Add contribution to morphism class. 
Extend functionality of morphism. 
Add morphism specializations in helena, spherical and cylindrical coordinates.

### CODE REVIEW

File: gyronimo/metrics/morphism.hh [hunk 1]
```diff
@@ -21,22 +21,60 @@
...
+	virtual dIR3 del(const IR3 &q) const = 0;
+
+	//! Returns the morphism's inverse derivatives, correspondent to the contravariant basis vectors in point `q`.
+	virtual dIR3 del_inverse(const IR3 &q) const;
+
+	virtual dIR3 tan_basis(const IR3& q) const {return del(q);};
+	virtual dIR3 dual_basis(const IR3& q) const {return del_inverse(q);};
+
+	// //! Returns the morphism's second derivatives, calculated in point `q`.
+	// virtual dSM3 del2(const IR3& q) const = 0;
```
Comment (prodrigs) [inline gyronimo/metrics/morphism.hh]: If not needed, these lines should be removed.

File: gyronimo/metrics/morphism.hh [hunk 2]
```diff
@@ -21,22 +21,60 @@
...
+	virtual dSM3 g_del(const IR3& q) const = 0; // to evade the problem of not getting a metric through the morphism (yet)
```
Comment (prodrigs) [inline gyronimo/metrics/morphism.hh]: This function g_del is a "metric" functionality already implemented, it should not be  repeated as part of a morphism. This leads to code redundancy and opens the way to possible mismatches.

File: gyronimo/metrics/morphism.hh (hunk 1 above)
Comment (tikoh-station) [reply gyronimo/metrics/morphism.hh]: Taking care of it.

File: gyronimo/metrics/morphism.hh (hunk 2 above)
Comment (tikoh-station) [reply gyronimo/metrics/morphism.hh]: I will review the relationship between `metric_covariant` and `morphism` to find a better way to do this.

File: gyronimo/core/IR3algebra.hh [hunk 3]
```diff
@@ -106,13 +106,40 @@ class dIR3 {
   const double& operator[](index i) const {return data_[i];}
   template <typename T>
   dIR3& operator=(const T& expr) {
-    for(size_t i = 0;i < 3;i++) data_[i] = expr[i];
+    for(size_t i = 0;i < 9;i++) data_[i] = expr[i];
```
Comment (prodrigs) [inline gyronimo/core/IR3algebra.hh:109]: Nicely spotted!

File: gyronimo/metrics/morphism_cylindrical.hh [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+// You should have received a copy of the GNU General Public License
+// along with ::gyronimo::.  If not, see <https://www.gnu.org/licenses/>.
+
+// @morphism_cylindrical.hh, this file is part of ::gyronimo::
+
+#ifndef GYRONIMO_MORPHISM_CYLINDRICAL
+#define GYRONIMO_MORPHISM_CYLINDRICAL
+
+#include <cmath>
```
Comment (prodrigs) [inline gyronimo/metrics/morphism_cylindrical.hh]: There are no cmath function calls in this file, the include directive should be moved to morphism_cylindrical.cc.

File: gyronimo/metrics/morphism_cylindrical.cc [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+		\begin{aligned}
+			r &= \sqrt{x^2 + y^2} \\
+			\phi &= \arctan \left( \frac{y}{x} \right) \\
+			z &= z
+		\end{aligned}
+    @f}
+*/
+IR3 morphism_cylindrical::inverse(const IR3 &x) const {
+	return {sqrt(x[IR3::u]*x[IR3::u] + x[IR3::v]*x[IR3::v]),
+			atan2(x[IR3::v], x[IR3::u]), x[IR3::w]};
```
Comment (prodrigs) [inline gyronimo/metrics/morphism_cylindrical.cc]: You should need a std:: qualifier here... Am I missing a using namespace std (which should be avoided, if possible) somewhere?

Comment (tikoh-station) [reply gyronimo/metrics/morphism_cylindrical.cc]: I did not write `using namespace std; ` anywhere. I don't think `std::sqrt` needs the qualifier, but I will add it anyway for the sake of consistency and to make sure it is using the right one.

File: gyronimo/metrics/morphism_cylindrical.hh (hunk 4 above)
Comment (tikoh-station) [reply gyronimo/metrics/morphism_cylindrical.hh]: I will make the change.

File: gyronimo/core/contraction.hh [hunk 6]
```diff
@@ -34,6 +34,8 @@ enum contraction_index {first, second, third};
 
 template<contraction_index>
 IR3 contraction(const dIR3& dA, const IR3& B);
+template<contraction_index, contraction_index>
+IR3 contraction(const ddIR3& ddA, const IR3& B, const IR3& C);
```
Comment (prodrigs) [inline gyronimo/core/contraction.hh:38]: Where are these contractions used? Are they needed in this pull request?

File: gyronimo/metrics/morphism_helena.hh [hunk 7]
```diff
@@ -20,27 +20,36 @@
 #ifndef GYRONIMO_MORPHISM_HELENA
 #define GYRONIMO_MORPHISM_HELENA
 
+#include <gsl/gsl_vector.h>
+#include <gsl/gsl_multiroots.h>
+#include <numbers>
+#include <gyronimo/core/error.hh>
+#include <gyronimo/core/multiroot.hh>
 #include <gyronimo/metrics/morphism.hh>
```
Comment (prodrigs) [inline gyronimo/metrics/morphism_helena.hh]: Why have all these new includes been added? If they are not explicitly needed in this include file, they should be removed.

File: gyronimo/metrics/morphism_helena.hh [hunk 8]
```diff
@@ -20,27 +20,36 @@
...
-  virtual IR3 operator()(const IR3& q) const override;
-  virtual IR3 inverse(const IR3& x) const override;
-  virtual dIR3 del(const IR3& q) const override;
-  const parser_helena* parser() const {return parser_;};
- private:
-  const parser_helena *parser_;
-  interpolator2d *R_, *z_;
-  static double reduce_2pi(double x);
-  static std::tuple<double, double> reflection_past_axis(double s, double chi);
+public:
```
Comment (prodrigs) [inline gyronimo/metrics/morphism_helena.hh:32]: Most of the replacements listed here are just the same lines but with a different number of spaces. So, there are almost no changes, but git lists a lot of them. This practice should be avoided in order to focus on the real changes.

File: gyronimo/metrics/morphism_helena.cc [hunk 9]
```diff
@@ -17,67 +17,111 @@
 
 // @morphism_helena.cc, this file is part of ::gyronimo::
 
-#include <numbers>
-#include <gyronimo/core/error.hh>
-#include <gyronimo/core/multiroot.hh>
 #include <gyronimo/metrics/morphism_helena.hh>
+#include <gsl/gsl_vector.h>
+#include <gsl/gsl_multiroots.h>
```
Comment (prodrigs) [inline gyronimo/metrics/morphism_helena.cc]: Low-level gsl operations should be handled by `multiroot` alone. Are these gsl includes really needed here?

File: gyronimo/metrics/morphism_helena.cc [hunk 10]
```diff
@@ -17,67 +17,111 @@
...
-  double Rgeo = p->rgeo();
-  double a = p->eps()*Rgeo;
-  dblock_adapter s_range(p->s()), chi_range(p->chi());
-  R_ = ifactory->interpolate_data(
-      s_range, chi_range, dblock_adapter(
-          parser_helena::narray_type(a*p->x() + Rgeo)));
-  z_ = ifactory->interpolate_data(
-      s_range, chi_range, dblock_adapter(
-          parser_helena::narray_type(a*p->y() + 0.0)));
+		const parser_helena *p, const interpolator2d_factory *ifactory)
```
Comment (prodrigs) [inline gyronimo/metrics/morphism_helena.cc:27]: Again, these changes are just blank spaces...

**Comment by prodrigs** [comment]
Variables `L0`, `r0`, and `radius_norm` are all used to normalise distances. Maybe the naming should be made uniform, e.g. `Lref`.

**Pull request merged**